File: NumericalDifferentiationIntegration/mid.cpp
#include <iostream>
#include <cmath>

#define pi 3.14159265358979323846

double f(double x) {
    return sin(x);
}

// Midpoint rule for numerical integration
double Midpoint(double (*f)(double x), double a,
                    double b, int N) {
    double dx = (b - a) / (N - 1);
    double x = a;

    double s = 0;
    for (int i = 0; i < N-1; ++i) {
        x = a + i*dx;
        s += f(x + 0.5*dx) * dx;
    }

    return s;
}


int main() {
    double a, b;
    std::cout << "[a b] = ";
    std::cin >> a >> b;
    int N;
    std::cout << "Number of discretized points: ";
    std::cin >> N;

    double result = Midpoint(f, a, b, N);
    std::cout << "Midpoint rule\n";
    std::cout << result << "\n";
    printf("%e\n", result);
  
    return 0;
}

File: NumericalDifferentiationIntegration/mid.py
from math import (pi, exp, sin, cos)

def f(x):
    return sin(x)

def Midpoint(f, a, b, N):
    """Midpoint rule for numerical integration"""
    dx = (b - a) / (N - 1)

    s = 0
    for i in range(N-1):
        x = a + i*dx
        s += f(x + 0.5*dx) * dx
    
    return s

a, b = map(eval, input("a, b = ").split(", "))
N = int(input("Number of discretized points: "))

print(Midpoint.__doc__)
print(Midpoint(f, a, b, N))
